### PR TITLE
feat: empty tree trash

### PR DESCRIPTION
**Pull request by Leeeon233**
This PR provides a new function called `empty_trash()` to clean up all caches under the `TRASH` node of the Tree Container.

We model the operation of deleting a node by moving the target node to a specific `TRASH` node, enabling concurrent moves of the target node to be applied as well.

All deleted nodes and their hierarchical relationships are cached even though they may never reappear. However, they still occupy memory. So we need gc them.

## Example

```ts
const loro = new Loro();
const tree = loro.getTree("root");
const root = tree.createNode();
const child = tree.createNode(root.id);
const child2 = tree.createNode(root.id);
tree.delete(child.id);
tree.delete(child2.id);
tree.emptyTrash();
loro.commit();
```

## Breaking Change

- A new Op has been added to the Tree Container, making the older version incompatible with this version. (But if no new Op is used, it does not affect the compatibility.

### CODE REVIEW

File: crates/loro-internal/src/encoding/value.rs [hunk 1]
```diff
@@ -170,6 +173,7 @@ pub enum Value<'a> {
 #[derive(Debug)]
 pub enum FutureValue<'a> {
     // The future value cannot depend on the arena for encoding.
+    EmptyTreeTrash(Arc<Vec<TreeID>>),
```
Comment (Leeeon233) [inline crates/loro-internal/src/encoding/value.rs:176]: Currently users must upgrade to be able to parse `EmptyTrash` Op, so can we put this Op in `Value` to reduce the encoding size?

File: crates/loro-internal/src/delta/tree.rs [hunk 2]
```diff
@@ -31,6 +32,7 @@ pub enum TreeExternalDiff {
         position: FractionalIndex,
     },
     Delete,
+    EmptyTrash,
```
Comment (Leeeon233) [inline crates/loro-internal/src/delta/tree.rs:35]: Should we show user which nodes have been emptied?

Comment (zxch3n) [reply crates/loro-internal/src/delta/tree.rs:35]: Can it be treated as Delete as well? In user land, we don't have to differentiate the two, should we?

File: crates/loro-internal/src/encoding/value.rs (hunk 1 above)
Comment (zxch3n) [reply crates/loro-internal/src/encoding/value.rs:176]: Do we have other options?

File: crates/loro-internal/src/delta/tree.rs (hunk 2 above)
Comment (Leeeon233) [reply crates/loro-internal/src/delta/tree.rs:35]: I don't think so. For users, `Delete` means delete the node, and `EmptyTrash` means **delete (clean)** the deleted node. We can't treat them the same.

Currently I only emit an event to tell the user that the `Trash` has been emptied.

File: crates/loro-internal/src/encoding/value.rs (hunk 1 above)
Comment (zxch3n) [reply crates/loro-internal/src/encoding/value.rs:176]: What are the possible designs and their trade-offs?

File: crates/loro-internal/src/delta/tree.rs (hunk 2 above)
Comment (zxch3n) [reply crates/loro-internal/src/delta/tree.rs:35]: But this is just an event. What are the differences from the user's perspective? It's confusing if users should handle them in the same way.

File: crates/loro-internal/src/encoding/value.rs (hunk 1 above)
Comment (Leeeon233) [reply crates/loro-internal/src/encoding/value.rs]: I moved it to `Value`

**Comment by zxch3n** [comment]
Does this PR include the tests for:
- JSON import/export schema (the related doc also needs updating)
- Forward & Backward Compatibility

File: crates/loro/src/lib.rs [hunk 3]
```diff
@@ -1461,6 +1461,22 @@ impl LoroTree {
...
+    /// We still keep the relevant information in the Op history, checkout to the any version will still work.
+    ///
+    /// ## Some Notes:
+    /// We recommend that you use this method with caution and only when you think you have enough historical operations.
+    /// Because if you empty the trash, and there are some concurrent operations that moving the deleted nodes back
+    /// to the tree. These operations may not have an effect as they have been completely removed from the State.
+    ///
+    /// To avoid unnecessary memory and storage usage for this Operation, you need to ensure that it is
+    /// only called once in potentially concurrent operations.
+    pub fn empty_trash(&self) -> LoroResult<()> {
```
Comment (zxch3n) [inline crates/loro/src/lib.rs:1476]: I think we need a param to allow users to only empty trash that is old enough. 

For example, by using the lamport value, or the timestamp, or allowing users to decide the maximal items allowed in the trash

Comment (Leeeon233) [reply crates/loro/src/lib.rs:1476]: Yes, it's better.

Comment (Leeeon233) [reply crates/loro/src/lib.rs]: Done.   For now, I just added `max_lamport`, if more parameters are needed, add them when necessary.

**Comment by Leeeon233** [comment]
closed by gc mode

**Pull request closed without merging**